src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs

```diff
@@ -949,27 +954,61 @@ private object GetValueFromSourceRow(int destRowIndex, out bool isSqlType, out b
                                 Debug.Assert(!(value is INullable) || !((INullable)value).IsNull, "IsDBNull returned false, but GetValue returned a null INullable");
                             }
 #endif
-                            return value;
+                            return WriteValueAsync(value, destRowIndex, isSqlType, isDataFeed, isNull);
                         }
                     }
                     else
                     {
-                        isDataFeed = false;
-
-                        IDataReader rowSourceAsIDataReader = (IDataReader)_rowSource;
+                        IDataReader r = (IDataReader)_rowSource;
 
                         // Only use IsDbNull when streaming is enabled and only for non-SqlDataReader
-                        if ((_enableStreaming) && (_SqlDataReaderRowSource == null) && (rowSourceAsIDataReader.IsDBNull(sourceOrdinal)))
+                        if ((_enableStreaming) && (_SqlDataReaderRowSource == null) && (r.IsDBNull(sourceOrdinal)))
                         {
-                            isSqlType = false;
                             isNull = true;
-                            return DBNull.Value;
+                            return WriteValueAsync(DBNull.Value, destRowIndex, isSqlType, isDataFeed, isNull);
                         }
                         else
                         {
-                            object columnValue = rowSourceAsIDataReader.GetValue(sourceOrdinal);
-                            ADP.IsNullOrSqlType(columnValue, out isNull, out isSqlType);
-                            return columnValue;
+                            if (_currentRowMetadata[sourceOrdinal].Method == ValueMethod.GetValue || r.IsDBNull(sourceOrdinal))
+                            {
+                                object columnValue = r.GetValue(sourceOrdinal);
+                                ADP.IsNullOrSqlType(columnValue, out isNull, out isSqlType);
+                                return WriteValueAsync(columnValue, destRowIndex, isSqlType, isDataFeed, isNull);
+                            }
+                            else
+                            {
+                                switch (_currentRowMetadata[sourceOrdinal].Method)
```

Comment:
would a bitwise enum improve performance of this `switch`?

Reply:
Probably not. Assuming that we don't have a set of known hot values we're going to do on average half the number of comparisons that there are values in the enum. The fastest way to do it is probably to manually produce a jump table for enum->func which will get you a linear performance character at the expense of an additional stack frame. Is it worth the effort and code complexity? You'd have to profile it and show that it's a decent percentage of the profile and I think it's unlikely.